src/tech/v3/dataset/impl/column_index_structure.clj

```diff
@@ -0,0 +1,85 @@
+(ns tech.v3.dataset.impl.column-index-structure
+  (:import [java.util TreeMap LinkedHashMap]
+           [tech.v3.datatype ListPersistentVector])
+  (:require [tech.v3.protocols.column :as col-proto]
+            [tech.v3.dataset.impl.column-base :refer [column-datatype-categorical?]]
+
+            [tech.v3.datatype :refer [elemwise-datatype clone ->buffer]]
+            [tech.v3.datatype.argops :refer [arggroup]]
+            [tech.v3.datatype.casting :refer [datatype->object-class]]
+            [clojure.set :refer [difference]]))
+
+
+(extend-type TreeMap
+  col-proto/PIndexStructure
+  (select-from-index [index-structure mode selection-spec]
+    (case mode
+      :pick
+      (let [^TreeMap new-index-structure (.clone ^TreeMap index-structure)
+            s (difference (set (.keySet new-index-structure)) (set selection-spec))]
+        (doseq [k s]
+          (.remove new-index-structure k))
+        new-index-structure)
```

Comment:
If you expect that selection spec is small as compared to the number of keys in index-structure it would be better to do a select-keys type operation.  I bet the average case here is that `selection-spec <<< key-set`.